In [261]:
## 라이브러리 로드
import pandas as pd

### 파일의 경로
- 절대 경로
    - 절대적인 주소 값
    - 환경이 변경되더라도 같은 위치를 지정
    - ex) c:/users/admin/a.txt, https://www.google.com
- 상대 경로 "
    - 상대적인 주소 값
    - 환경이 변경되었을 때 위치도 변경
    - ex)
        - ./ : 현재 작업 중인 디렉토리
        - ../ : 현재 디렉토리에서 상위 폴더로 이동
        - 폴더명/ : 해당 디렉토리로 이동
    - 절대 경로보다 상대 경로를 쓰는 경우가 더 많다.

### 외부의 데이터 파일을 로드
- csv
    - ','로 이루어진 데이터
    - read_csv()
- jason
    - jason 데이터는 dict와 흡사
    - read_json()
- excel
    - read_excel()
    - xls, xlsx 파일 확장자에 따라서 추가적인 라이브러리 설치
- xml
    - read_xml()
    

In [262]:
## 외부의 csv 파일을 로드
## 상대 경로
## 상위 폴더로 이동(../) -> csv라는 하위 폴더로 이동(csv/) -> corona.csv
df = pd.read_csv("../data/csv/corona.csv") #csv는 소문자니까 함수겠구나~ 추측 가능. #경로 설정시 따옴표 필요
df

Unnamed: 0                 createDt  deathCnt  decideCnt  seq   stateDt  \
0             0  2022-06-08 09:09:05.982     24305   18188200  904  20220608   
1             1  2022-06-07 09:09:00.897     24299   18174842  903  20220607   
2             2  2022-06-06 09:00:06.734     24279   18168670  902  20220606   
3             3  2022-06-05 08:53:19.426     24258   18163648  901  20220605   
4             4  2022-06-04 08:56:49.219     24238   18153814  900  20220604   
..          ...                      ...       ...        ...  ...       ...   
815         815  2020-03-14 00:00:00.000        72       8086   55  20200314   
816         816  2020-03-13 00:00:00.000        67       7979   54  20200313   
817         817  2020-03-12 00:00:00.000        66       7869   53  20200312   
818         818  2020-03-11 00:00:00.000        60       7755   52  20200311   
819         819  2020-03-10 00:00:00.000        54       7513   51  20200310   

    stateTime                 updateDt  accExamCnt  accDefRate  
0       00:00                      NaN         NaN         NaN  
1       00:00  2022-06-08 09:10:36.846         NaN         NaN  
2       00:00  2022-06-08 09:10:50.441         NaN         NaN  
3       00:00  2022-06-08 09:11:04.758         NaN         NaN  
4       00:00  2022-06-08 09:11:26.303         NaN         NaN  
..        ...                      ...         ...         ...  
815     00:00   2021-10-07 10:30:51.51    261335.0    3.318000  
816     00:00   2021-10-07 10:30:51.51    248647.0    3.458499  
817     00:00   2021-10-07 10:30:51.51    234998.0    3.621744  
818     00:00   2021-10-07 10:30:51.51    222395.0    3.804175  
819     00:00   2021-10-07 10:30:51.51    210144.0    3.919308  

[820 rows x 10 columns]

In [263]:
## 절대 경로 사용
## 좌측 패널에서 해당 파일 우클릭 > "경로 복사" or "상대 경로 복사" 클릭
## 다만 역슬러시가 2번씩 나와야 하므로, 각 경로마다 역슬러시를 추가해줘야 함.
## 재택근무, 외장하드 활용 등의 변수가 있으므로 절대 경로보다는 상대 경로 사용을 추천!
df2 = pd.read_csv("D:\\240120_olA1ol\\learning_data\\data\\csv\\corona.csv")
df2

Unnamed: 0                 createDt  deathCnt  decideCnt  seq   stateDt  \
0             0  2022-06-08 09:09:05.982     24305   18188200  904  20220608   
1             1  2022-06-07 09:09:00.897     24299   18174842  903  20220607   
2             2  2022-06-06 09:00:06.734     24279   18168670  902  20220606   
3             3  2022-06-05 08:53:19.426     24258   18163648  901  20220605   
4             4  2022-06-04 08:56:49.219     24238   18153814  900  20220604   
..          ...                      ...       ...        ...  ...       ...   
815         815  2020-03-14 00:00:00.000        72       8086   55  20200314   
816         816  2020-03-13 00:00:00.000        67       7979   54  20200313   
817         817  2020-03-12 00:00:00.000        66       7869   53  20200312   
818         818  2020-03-11 00:00:00.000        60       7755   52  20200311   
819         819  2020-03-10 00:00:00.000        54       7513   51  20200310   

    stateTime                 updateDt  accExamCnt  accDefRate  
0       00:00                      NaN         NaN         NaN  
1       00:00  2022-06-08 09:10:36.846         NaN         NaN  
2       00:00  2022-06-08 09:10:50.441         NaN         NaN  
3       00:00  2022-06-08 09:11:04.758         NaN         NaN  
4       00:00  2022-06-08 09:11:26.303         NaN         NaN  
..        ...                      ...         ...         ...  
815     00:00   2021-10-07 10:30:51.51    261335.0    3.318000  
816     00:00   2021-10-07 10:30:51.51    248647.0    3.458499  
817     00:00   2021-10-07 10:30:51.51    234998.0    3.621744  
818     00:00   2021-10-07 10:30:51.51    222395.0    3.804175  
819     00:00   2021-10-07 10:30:51.51    210144.0    3.919308  

[820 rows x 10 columns]

In [264]:
# 데이터프레임의 상위 데이터를 추출 (값 안 쓰면 5개)
df.head(2)

# 결측치 NaN이 생기는 이유.. 수집을 더 이상 하지 않기 때문일 수도 있다.

Unnamed: 0                 createDt  deathCnt  decideCnt  seq   stateDt  \
0           0  2022-06-08 09:09:05.982     24305   18188200  904  20220608   
1           1  2022-06-07 09:09:00.897     24299   18174842  903  20220607   

  stateTime                 updateDt  accExamCnt  accDefRate  
0     00:00                      NaN         NaN         NaN  
1     00:00  2022-06-08 09:10:36.846         NaN         NaN

In [265]:
# 데이터프레임의 하위 데이터를 추출 (값 안 쓰면 5개)

df.tail()

Unnamed: 0                 createDt  deathCnt  decideCnt  seq   stateDt  \
815         815  2020-03-14 00:00:00.000        72       8086   55  20200314   
816         816  2020-03-13 00:00:00.000        67       7979   54  20200313   
817         817  2020-03-12 00:00:00.000        66       7869   53  20200312   
818         818  2020-03-11 00:00:00.000        60       7755   52  20200311   
819         819  2020-03-10 00:00:00.000        54       7513   51  20200310   

    stateTime                updateDt  accExamCnt  accDefRate  
815     00:00  2021-10-07 10:30:51.51    261335.0    3.318000  
816     00:00  2021-10-07 10:30:51.51    248647.0    3.458499  
817     00:00  2021-10-07 10:30:51.51    234998.0    3.621744  
818     00:00  2021-10-07 10:30:51.51    222395.0    3.804175  
819     00:00  2021-10-07 10:30:51.51    210144.0    3.919308

In [266]:
# 데이터프레임의 정보를 출력
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 820 entries, 0 to 819
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  820 non-null    int64  
 1   createDt    820 non-null    object 
 2   deathCnt    820 non-null    int64  
 3   decideCnt   820 non-null    int64  
 4   seq         820 non-null    int64  
 5   stateDt     820 non-null    int64  
 6   stateTime   820 non-null    object 
 7   updateDt    817 non-null    object 
 8   accExamCnt  692 non-null    float64
 9   accDefRate  623 non-null    float64
dtypes: float64(2), int64(5), object(3)
memory usage: 64.2+ KB


int = 정수 형태
float = 실수 형태 (소수점 있는)
### int보다는 float 형태를 더 선호한다. 연산 처리가 float이 더 빠르고 정확하다고 한다.

정수형 (int), 실수형 (float), 논리형 (bool), 문자열 (string) 등이 있다.

null값은 매우 중요함. 

"결측치"를 제거할 수도 있고, 특정 값으로 채워줄 수도 있다.
분석할 때 선형 대수.. 등등을 도입할 때
데이터의 편향성이 생길 수 있는데
분석값 개개인의 스펙이 된다..
채우는 방법이 워낙 다양한데 
편향성을 일으키기 매우 쉬움.
데이터에 어떤 값을 집어넣는가에 따라 달라지는 거..

결측치를 제거한다는 건 5% 내외.
3%로 잡는 분도 있고 5%로 잡는 분도 있다.
####### 결측치 = NaN
엑셀로 치면 공란에 해당한다.

In [267]:
## 코드에서 결측치를 확인하는 방법.
## Na가 존재하는가 존재하지 않는가? 확인하는 함수를 사용.
## isna()

df.isna()

Unnamed: 0  createDt  deathCnt  decideCnt    seq  stateDt  stateTime  \
0         False     False     False      False  False    False      False   
1         False     False     False      False  False    False      False   
2         False     False     False      False  False    False      False   
3         False     False     False      False  False    False      False   
4         False     False     False      False  False    False      False   
..          ...       ...       ...        ...    ...      ...        ...   
815       False     False     False      False  False    False      False   
816       False     False     False      False  False    False      False   
817       False     False     False      False  False    False      False   
818       False     False     False      False  False    False      False   
819       False     False     False      False  False    False      False   

     updateDt  accExamCnt  accDefRate  
0        True        True        True  
1       False        True        True  
2       False        True        True  
3       False        True        True  
4       False        True        True  
..        ...         ...         ...  
815     False       False       False  
816     False       False       False  
817     False       False       False  
818     False       False       False  
819     False       False       False  

[820 rows x 10 columns]

In [268]:
## 결측치가 아닌 데이터는 false로 나오고, 결측치인 데이터는 true로 집계
## false = 0, true = 1
## 모든 값을 더해줬을 때, 총 몇 개의 결측치가 있는지.. 살펴볼 수 있다.

In [269]:
## 데이터프레임에서 결측치 개수를 확인
df.isna().sum()

Unnamed: 0      0
createDt        0
deathCnt        0
decideCnt       0
seq             0
stateDt         0
stateTime       0
updateDt        3
accExamCnt    128
accDefRate    197
dtype: int64

In [270]:
## 해당하는 칼럼별로 더해준다는 의미.
## 마지막 3개 라인에서.. 각 칼럼에 결측치가 몇 개씩 존재하는지. 알 수 있다.
## isna만 단독적으로 사용할 수도 있고, sum을 이용해서 그 전체 개수를 파악할 수도 있다.

In [271]:
a = df.isna()

In [272]:
a.sum()

## 요 2개 함수를 합친 형태인 것.

Unnamed: 0      0
createDt        0
deathCnt        0
decideCnt       0
seq             0
stateDt         0
stateTime       0
updateDt        3
accExamCnt    128
accDefRate    197
dtype: int64

In [273]:
a.sum(axis=1)

0      3
1      2
2      2
3      2
4      2
      ..
815    0
816    0
817    0
818    0
819    0
Length: 820, dtype: int64

In [274]:
## 결측치를 제거: drop을 쓴다.
# dropna(axis = {0('rows') | 1('columns')})

# 행을 기준으로 결측치를 제거
df.dropna(axis=0)

Unnamed: 0                 createDt  deathCnt  decideCnt  seq   stateDt  \
197         197  2021-11-23 08:59:47.863      3328     420946  706  20211123   
198         198  2021-11-22 09:07:45.615      3298     418249  705  20211122   
199         199  2021-11-21 09:05:22.138      3274     415422  704  20211121   
200         200  2021-11-20 08:51:26.319      3244     412302  703  20211120   
201         201  2021-11-19 08:54:43.169      3215     409097  702  20211119   
..          ...                      ...       ...        ...  ...       ...   
815         815  2020-03-14 00:00:00.000        72       8086   55  20200314   
816         816  2020-03-13 00:00:00.000        67       7979   54  20200313   
817         817  2020-03-12 00:00:00.000        66       7869   53  20200312   
818         818  2020-03-11 00:00:00.000        60       7755   52  20200311   
819         819  2020-03-10 00:00:00.000        54       7513   51  20200310   

    stateTime                 updateDt  accExamCnt  accDefRate  
197     00:00  2021-11-30 09:06:13.348  16878955.0    2.787908  
198     00:00  2021-11-26 09:29:38.144  16816616.0    2.778249  
199     00:00  2021-11-26 09:29:24.197  16777679.0    2.760823  
200     00:00  2021-11-26 09:28:52.855  16740119.0    2.740178  
201     00:00  2021-11-26 09:28:44.079  16684168.0    2.723426  
..        ...                      ...         ...         ...  
815     00:00   2021-10-07 10:30:51.51    261335.0    3.318000  
816     00:00   2021-10-07 10:30:51.51    248647.0    3.458499  
817     00:00   2021-10-07 10:30:51.51    234998.0    3.621744  
818     00:00   2021-10-07 10:30:51.51    222395.0    3.804175  
819     00:00   2021-10-07 10:30:51.51    210144.0    3.919308  

[623 rows x 10 columns]

In [275]:
# 데이터가 0~196번이 다 지워져서, 623개의 데이터만 남게 됨.
# 결측치 제거 = 해당하는 데이터를 모두 날려버린다..
# 결측치를 채워주는 부분도 있다..
# 데이터 개수가 몇 개 안 되면 지워지는 게 맞을 수 있고 (5% 미만)
# 지울지 말지를 잘 봐야 한다..

In [276]:
# 열을 기준으로 결측치를 제거
df.dropna(axis='columns') #1이라고 적어도, COLUMNS라고 적어도 됨
#df는 변화하지 않음. 원본 파일, 데이터프레임 모두 그대로 있다. 출력값만 수정된 형태로 보이는 것.

Unnamed: 0                 createDt  deathCnt  decideCnt  seq   stateDt  \
0             0  2022-06-08 09:09:05.982     24305   18188200  904  20220608   
1             1  2022-06-07 09:09:00.897     24299   18174842  903  20220607   
2             2  2022-06-06 09:00:06.734     24279   18168670  902  20220606   
3             3  2022-06-05 08:53:19.426     24258   18163648  901  20220605   
4             4  2022-06-04 08:56:49.219     24238   18153814  900  20220604   
..          ...                      ...       ...        ...  ...       ...   
815         815  2020-03-14 00:00:00.000        72       8086   55  20200314   
816         816  2020-03-13 00:00:00.000        67       7979   54  20200313   
817         817  2020-03-12 00:00:00.000        66       7869   53  20200312   
818         818  2020-03-11 00:00:00.000        60       7755   52  20200311   
819         819  2020-03-10 00:00:00.000        54       7513   51  20200310   

    stateTime  
0       00:00  
1       00:00  
2       00:00  
3       00:00  
4       00:00  
..        ...  
815     00:00  
816     00:00  
817     00:00  
818     00:00  
819     00:00  

[820 rows x 7 columns]

In [277]:
## 결측치를 특정한 데이터로 치환
# fillna - 함수 이름은 직관적.
# fillna({data}) : 특정한 데이터로 결측치를 치환
# fillna(method = {'ffill' | 'bfill'}) 가장 많이 쓰는 옵션이 ffill 또는 bfill
# ffill은 forward의 앞, bfill은 backward의 뒤
# 앞에 있는 데이터를 끌어와서 복사시켜 지워버리겠다
# 뒤에 있는 데이터로 치환하겠다

df.fillna(10)

Unnamed: 0                 createDt  deathCnt  decideCnt  seq   stateDt  \
0             0  2022-06-08 09:09:05.982     24305   18188200  904  20220608   
1             1  2022-06-07 09:09:00.897     24299   18174842  903  20220607   
2             2  2022-06-06 09:00:06.734     24279   18168670  902  20220606   
3             3  2022-06-05 08:53:19.426     24258   18163648  901  20220605   
4             4  2022-06-04 08:56:49.219     24238   18153814  900  20220604   
..          ...                      ...       ...        ...  ...       ...   
815         815  2020-03-14 00:00:00.000        72       8086   55  20200314   
816         816  2020-03-13 00:00:00.000        67       7979   54  20200313   
817         817  2020-03-12 00:00:00.000        66       7869   53  20200312   
818         818  2020-03-11 00:00:00.000        60       7755   52  20200311   
819         819  2020-03-10 00:00:00.000        54       7513   51  20200310   

    stateTime                 updateDt  accExamCnt  accDefRate  
0       00:00                       10        10.0   10.000000  
1       00:00  2022-06-08 09:10:36.846        10.0   10.000000  
2       00:00  2022-06-08 09:10:50.441        10.0   10.000000  
3       00:00  2022-06-08 09:11:04.758        10.0   10.000000  
4       00:00  2022-06-08 09:11:26.303        10.0   10.000000  
..        ...                      ...         ...         ...  
815     00:00   2021-10-07 10:30:51.51    261335.0    3.318000  
816     00:00   2021-10-07 10:30:51.51    248647.0    3.458499  
817     00:00   2021-10-07 10:30:51.51    234998.0    3.621744  
818     00:00   2021-10-07 10:30:51.51    222395.0    3.804175  
819     00:00   2021-10-07 10:30:51.51    210144.0    3.919308  

[820 rows x 10 columns]

In [278]:
df.fillna(10).head(5)

#맨 앞 5개 데이터에서 결측치를 10으로 채워라.
#문자로 채울 수도 있지만 본 데이터들이 float 형태이므로 10으로 집어넣어본 것.
#updateDt는 날짜&시간이어서 문자로 인식한 것이고 나머지는 float..

Unnamed: 0                 createDt  deathCnt  decideCnt  seq   stateDt  \
0           0  2022-06-08 09:09:05.982     24305   18188200  904  20220608   
1           1  2022-06-07 09:09:00.897     24299   18174842  903  20220607   
2           2  2022-06-06 09:00:06.734     24279   18168670  902  20220606   
3           3  2022-06-05 08:53:19.426     24258   18163648  901  20220605   
4           4  2022-06-04 08:56:49.219     24238   18153814  900  20220604   

  stateTime                 updateDt  accExamCnt  accDefRate  
0     00:00                       10        10.0        10.0  
1     00:00  2022-06-08 09:10:36.846        10.0        10.0  
2     00:00  2022-06-08 09:10:50.441        10.0        10.0  
3     00:00  2022-06-08 09:11:04.758        10.0        10.0  
4     00:00  2022-06-08 09:11:26.303        10.0        10.0

In [279]:
df.fillna('a').head(5)

Unnamed: 0                 createDt  deathCnt  decideCnt  seq   stateDt  \
0           0  2022-06-08 09:09:05.982     24305   18188200  904  20220608   
1           1  2022-06-07 09:09:00.897     24299   18174842  903  20220607   
2           2  2022-06-06 09:00:06.734     24279   18168670  902  20220606   
3           3  2022-06-05 08:53:19.426     24258   18163648  901  20220605   
4           4  2022-06-04 08:56:49.219     24238   18153814  900  20220604   

  stateTime                 updateDt accExamCnt accDefRate  
0     00:00                        a          a          a  
1     00:00  2022-06-08 09:10:36.846          a          a  
2     00:00  2022-06-08 09:10:50.441          a          a  
3     00:00  2022-06-08 09:11:04.758          a          a  
4     00:00  2022-06-08 09:11:26.303          a          a

In [280]:
df.fillna('a').info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 820 entries, 0 to 819
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  820 non-null    int64 
 1   createDt    820 non-null    object
 2   deathCnt    820 non-null    int64 
 3   decideCnt   820 non-null    int64 
 4   seq         820 non-null    int64 
 5   stateDt     820 non-null    int64 
 6   stateTime   820 non-null    object
 7   updateDt    820 non-null    object
 8   accExamCnt  820 non-null    object
 9   accDefRate  820 non-null    object
dtypes: int64(5), object(5)
memory usage: 64.2+ KB


In [281]:
df.fillna(method='ffill').head()

#앞에 있는 데이터를 끌어와서 결측치를 채우려고 했는데, 그 앞 데이터가 이미 결측치어서 앞에 걸 끌어와서 채울 수가 없다.

C:\Users\EZEN\AppData\Local\Temp\ipykernel_1784\1563425071.py:1: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df.fillna(method='ffill').head()


Unnamed: 0                 createDt  deathCnt  decideCnt  seq   stateDt  \
0           0  2022-06-08 09:09:05.982     24305   18188200  904  20220608   
1           1  2022-06-07 09:09:00.897     24299   18174842  903  20220607   
2           2  2022-06-06 09:00:06.734     24279   18168670  902  20220606   
3           3  2022-06-05 08:53:19.426     24258   18163648  901  20220605   
4           4  2022-06-04 08:56:49.219     24238   18153814  900  20220604   

  stateTime                 updateDt  accExamCnt  accDefRate  
0     00:00                      NaN         NaN         NaN  
1     00:00  2022-06-08 09:10:36.846         NaN         NaN  
2     00:00  2022-06-08 09:10:50.441         NaN         NaN  
3     00:00  2022-06-08 09:11:04.758         NaN         NaN  
4     00:00  2022-06-08 09:11:26.303         NaN         NaN

In [282]:
df.fillna(method='bfill').head()

C:\Users\EZEN\AppData\Local\Temp\ipykernel_1784\1592124609.py:1: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df.fillna(method='bfill').head()


Unnamed: 0                 createDt  deathCnt  decideCnt  seq   stateDt  \
0           0  2022-06-08 09:09:05.982     24305   18188200  904  20220608   
1           1  2022-06-07 09:09:00.897     24299   18174842  903  20220607   
2           2  2022-06-06 09:00:06.734     24279   18168670  902  20220606   
3           3  2022-06-05 08:53:19.426     24258   18163648  901  20220605   
4           4  2022-06-04 08:56:49.219     24238   18153814  900  20220604   

  stateTime                 updateDt  accExamCnt  accDefRate  
0     00:00  2022-06-08 09:10:36.846  21518073.0    2.787908  
1     00:00  2022-06-08 09:10:36.846  21518073.0    2.787908  
2     00:00  2022-06-08 09:10:50.441  21518073.0    2.787908  
3     00:00  2022-06-08 09:11:04.758  21518073.0    2.787908  
4     00:00  2022-06-08 09:11:26.303  21518073.0    2.787908

In [283]:
## 데이터프레임의 통계요약정보
df.describe()

Unnamed: 0      deathCnt     decideCnt         seq       stateDt  \
count  820.000000    820.000000  8.200000e+02  820.000000  8.200000e+02   
mean   409.500000   3927.832927  1.913097e+06  472.393902  2.020897e+07   
std    236.857904   6164.310693  4.857030e+06  249.701095  7.118265e+03   
min      0.000000     54.000000  7.513000e+03   51.000000  2.020031e+07   
25%    204.750000    415.750000  2.393525e+04  256.750000  2.020100e+07   
50%    409.500000   1812.500000  1.185640e+05  461.500000  2.021042e+07   
75%    614.250000   3120.500000  3.979915e+05  698.250000  2.021112e+07   
max    819.000000  24305.000000  1.818820e+07  904.000000  2.022061e+07   

         accExamCnt  accDefRate  
count  6.920000e+02  623.000000  
mean   7.545057e+06    1.556435  
std    6.172020e+06    0.543014  
min    2.101440e+05    0.902205  
25%    1.934309e+06    1.078089  
50%    6.368310e+06    1.416159  
75%    1.216890e+07    1.816009  
max    2.151807e+07    3.919308

### 데이터프레임 필터
조건을 기준으로 데이터를 볼 것이냐
위치를 기준으로 데이터를 볼 것이냐
두 가지로 설정할 수가 있음.

1. loc[{인덱스의 조건}, {컬럼의 조건}] (location)
    - 인덱스의 값을 기준으로 데이터프레임 필터
    - 컬럼의 값을 기준으로 데이터프레임 필터
    - 인덱스와 컬럼의 조건식을 이용하여 데이터프레임 필터
2. iloc[{인덱스의 위치}, {컬럼의 위치}] (integer location)
    - 인덱스나 컬럼의 위치를 기준으로 데이터프레임 필터

In [284]:
# 데이터프레임에서 특정한 데이터를 제거
# drop( {조건} , axis = {0(rows) | 1(columns)}, inplace = {bool} )
# 데이터프레임명.drop()
# axis : 기준 -> 행{0}을 지울 것이냐?  열{1}을 지울 것이냐?
# inplace : 참 / 거짓 형태의 데이터를 입력
# 참 : 기준이 되는 데이터프레임을 변경
# 거짓 : 기준이 되는 데이터프레임을 변경하지 않고 결과를 출력 (기본값)

# loc와 iloc의 차이를 보여주기 위한.. 부가 조건들

# df에서 인덱스 0인 데이터를 제거
# df에서 Unnamed: 0 컬럼을 제거 - "콜론하고 한 칸 띄고 0"
df.drop('Unnamed: 0', axis=1)
# 해당 컬럼의 이름이 없을 때 자동으로 Unnamed가 됨..

createDt  deathCnt  decideCnt  seq   stateDt stateTime  \
0    2022-06-08 09:09:05.982     24305   18188200  904  20220608     00:00   
1    2022-06-07 09:09:00.897     24299   18174842  903  20220607     00:00   
2    2022-06-06 09:00:06.734     24279   18168670  902  20220606     00:00   
3    2022-06-05 08:53:19.426     24258   18163648  901  20220605     00:00   
4    2022-06-04 08:56:49.219     24238   18153814  900  20220604     00:00   
..                       ...       ...        ...  ...       ...       ...   
815  2020-03-14 00:00:00.000        72       8086   55  20200314     00:00   
816  2020-03-13 00:00:00.000        67       7979   54  20200313     00:00   
817  2020-03-12 00:00:00.000        66       7869   53  20200312     00:00   
818  2020-03-11 00:00:00.000        60       7755   52  20200311     00:00   
819  2020-03-10 00:00:00.000        54       7513   51  20200310     00:00   

                    updateDt  accExamCnt  accDefRate  
0                        NaN         NaN         NaN  
1    2022-06-08 09:10:36.846         NaN         NaN  
2    2022-06-08 09:10:50.441         NaN         NaN  
3    2022-06-08 09:11:04.758         NaN         NaN  
4    2022-06-08 09:11:26.303         NaN         NaN  
..                       ...         ...         ...  
815   2021-10-07 10:30:51.51    261335.0    3.318000  
816   2021-10-07 10:30:51.51    248647.0    3.458499  
817   2021-10-07 10:30:51.51    234998.0    3.621744  
818   2021-10-07 10:30:51.51    222395.0    3.804175  
819   2021-10-07 10:30:51.51    210144.0    3.919308  

[820 rows x 9 columns]

In [285]:
df

Unnamed: 0                 createDt  deathCnt  decideCnt  seq   stateDt  \
0             0  2022-06-08 09:09:05.982     24305   18188200  904  20220608   
1             1  2022-06-07 09:09:00.897     24299   18174842  903  20220607   
2             2  2022-06-06 09:00:06.734     24279   18168670  902  20220606   
3             3  2022-06-05 08:53:19.426     24258   18163648  901  20220605   
4             4  2022-06-04 08:56:49.219     24238   18153814  900  20220604   
..          ...                      ...       ...        ...  ...       ...   
815         815  2020-03-14 00:00:00.000        72       8086   55  20200314   
816         816  2020-03-13 00:00:00.000        67       7979   54  20200313   
817         817  2020-03-12 00:00:00.000        66       7869   53  20200312   
818         818  2020-03-11 00:00:00.000        60       7755   52  20200311   
819         819  2020-03-10 00:00:00.000        54       7513   51  20200310   

    stateTime                 updateDt  accExamCnt  accDefRate  
0       00:00                      NaN         NaN         NaN  
1       00:00  2022-06-08 09:10:36.846         NaN         NaN  
2       00:00  2022-06-08 09:10:50.441         NaN         NaN  
3       00:00  2022-06-08 09:11:04.758         NaN         NaN  
4       00:00  2022-06-08 09:11:26.303         NaN         NaN  
..        ...                      ...         ...         ...  
815     00:00   2021-10-07 10:30:51.51    261335.0    3.318000  
816     00:00   2021-10-07 10:30:51.51    248647.0    3.458499  
817     00:00   2021-10-07 10:30:51.51    234998.0    3.621744  
818     00:00   2021-10-07 10:30:51.51    222395.0    3.804175  
819     00:00   2021-10-07 10:30:51.51    210144.0    3.919308  

[820 rows x 10 columns]

In [286]:
# df = df.drop('Unnamed: 0', axis=1)
df.drop('Unnamed: 0', axis=1, inplace=True)
#inplace는 기본값이 false.. 

In [287]:
df

createDt  deathCnt  decideCnt  seq   stateDt stateTime  \
0    2022-06-08 09:09:05.982     24305   18188200  904  20220608     00:00   
1    2022-06-07 09:09:00.897     24299   18174842  903  20220607     00:00   
2    2022-06-06 09:00:06.734     24279   18168670  902  20220606     00:00   
3    2022-06-05 08:53:19.426     24258   18163648  901  20220605     00:00   
4    2022-06-04 08:56:49.219     24238   18153814  900  20220604     00:00   
..                       ...       ...        ...  ...       ...       ...   
815  2020-03-14 00:00:00.000        72       8086   55  20200314     00:00   
816  2020-03-13 00:00:00.000        67       7979   54  20200313     00:00   
817  2020-03-12 00:00:00.000        66       7869   53  20200312     00:00   
818  2020-03-11 00:00:00.000        60       7755   52  20200311     00:00   
819  2020-03-10 00:00:00.000        54       7513   51  20200310     00:00   

                    updateDt  accExamCnt  accDefRate  
0                        NaN         NaN         NaN  
1    2022-06-08 09:10:36.846         NaN         NaN  
2    2022-06-08 09:10:50.441         NaN         NaN  
3    2022-06-08 09:11:04.758         NaN         NaN  
4    2022-06-08 09:11:26.303         NaN         NaN  
..                       ...         ...         ...  
815   2021-10-07 10:30:51.51    261335.0    3.318000  
816   2021-10-07 10:30:51.51    248647.0    3.458499  
817   2021-10-07 10:30:51.51    234998.0    3.621744  
818   2021-10-07 10:30:51.51    222395.0    3.804175  
819   2021-10-07 10:30:51.51    210144.0    3.919308  

[820 rows x 9 columns]

In [288]:
df.head()

createDt  deathCnt  decideCnt  seq   stateDt stateTime  \
0  2022-06-08 09:09:05.982     24305   18188200  904  20220608     00:00   
1  2022-06-07 09:09:00.897     24299   18174842  903  20220607     00:00   
2  2022-06-06 09:00:06.734     24279   18168670  902  20220606     00:00   
3  2022-06-05 08:53:19.426     24258   18163648  901  20220605     00:00   
4  2022-06-04 08:56:49.219     24238   18153814  900  20220604     00:00   

                  updateDt  accExamCnt  accDefRate  
0                      NaN         NaN         NaN  
1  2022-06-08 09:10:36.846         NaN         NaN  
2  2022-06-08 09:10:50.441         NaN         NaN  
3  2022-06-08 09:11:04.758         NaN         NaN  
4  2022-06-08 09:11:26.303         NaN         NaN

In [289]:
#뭔가를 실행했을 때 나오면, 데이터에는 변함이 없다는 걸로 이해하면 됨.
# 파이썬은 문법이 쉽고, 환경이 잘 돼 있다.. #시간 단축이 훨씬 될 것.

In [290]:
df2 = pd.read_csv("../data/csv/corona.csv")

In [291]:
df2 = df.copy() #실수 방지를 위해 중간 중간 복사본 만들어주면 됨.

In [292]:
df2 = df2.drop(0, axis=0)
# df2 = df2.drop(0, axis=0, inplace = True)

In [293]:
# df2

In [294]:
## 인덱스의 값이 1인 데이터를 출력
df2.loc[[1], ]

createDt  deathCnt  decideCnt  seq   stateDt stateTime  \
1  2022-06-07 09:09:00.897     24299   18174842  903  20220607     00:00   

                  updateDt  accExamCnt  accDefRate  
1  2022-06-08 09:10:36.846         NaN         NaN

In [295]:
df2.iloc[[1], ] # 두 번째 위치에 있는 데이터를 출력하겠다.

createDt  deathCnt  decideCnt  seq   stateDt stateTime  \
2  2022-06-06 09:00:06.734     24279   18168670  902  20220606     00:00   

                  updateDt  accExamCnt  accDefRate  
2  2022-06-08 09:10:50.441         NaN         NaN

In [296]:
## 특정 컬럼의 데이터만 출력
# df2.loc[,"createDt"]

In [297]:
#뒤와 앞
#앞에 안 넣어줬을 때 에러가 나는 경우
# - 앞 index는 전체를 본다 = 콜론을 하나 써주면 된다.
# 콜론 하나 = 맨 처음부터 맨 끝까지 보겠다는 의미.
df2.loc[ : , "createDt"]

1      2022-06-07 09:09:00.897
2      2022-06-06 09:00:06.734
3      2022-06-05 08:53:19.426
4      2022-06-04 08:56:49.219
5      2022-06-03 09:08:18.729
                ...           
815    2020-03-14 00:00:00.000
816    2020-03-13 00:00:00.000
817    2020-03-12 00:00:00.000
818    2020-03-11 00:00:00.000
819    2020-03-10 00:00:00.000
Name: createDt, Length: 819, dtype: object

In [298]:
# 스리즈 형태로 나타남
# 인덱스는 전체로 선택했지만, 컬럼은 하나만 선택했기 때문..
#여러 개의 문자가 들어갈 수 있는 타입으로 바꿔주려면 []를 넣어주면 된다.
df2.loc[ : , ["createDt"] ]

createDt
1    2022-06-07 09:09:00.897
2    2022-06-06 09:00:06.734
3    2022-06-05 08:53:19.426
4    2022-06-04 08:56:49.219
5    2022-06-03 09:08:18.729
..                       ...
815  2020-03-14 00:00:00.000
816  2020-03-13 00:00:00.000
817  2020-03-12 00:00:00.000
818  2020-03-11 00:00:00.000
819  2020-03-10 00:00:00.000

[819 rows x 1 columns]

In [299]:
df2[ ['createDt'] ] #이렇게 적어도 같은 결과가 나온다.
# 대괄호만 쓸 때는 column을 기준으로만 나온다.

createDt
1    2022-06-07 09:09:00.897
2    2022-06-06 09:00:06.734
3    2022-06-05 08:53:19.426
4    2022-06-04 08:56:49.219
5    2022-06-03 09:08:18.729
..                       ...
815  2020-03-14 00:00:00.000
816  2020-03-13 00:00:00.000
817  2020-03-12 00:00:00.000
818  2020-03-11 00:00:00.000
819  2020-03-10 00:00:00.000

[819 rows x 1 columns]

In [300]:
df2.iloc[ : , [0] ]

createDt
1    2022-06-07 09:09:00.897
2    2022-06-06 09:00:06.734
3    2022-06-05 08:53:19.426
4    2022-06-04 08:56:49.219
5    2022-06-03 09:08:18.729
..                       ...
815  2020-03-14 00:00:00.000
816  2020-03-13 00:00:00.000
817  2020-03-12 00:00:00.000
818  2020-03-11 00:00:00.000
819  2020-03-10 00:00:00.000

[819 rows x 1 columns]

In [301]:
df2.iloc[ : , 1:3] #column 두 번째 위치부터 4번째 위치 직전까지 보겠다.
#iloc는 마지막 지정자 값을 포함시키지 않고, loc는 마지막 지정자 값을 포함시킨다.

deathCnt  decideCnt
1       24299   18174842
2       24279   18168670
3       24258   18163648
4       24238   18153814
5       24229   18141775
..        ...        ...
815        72       8086
816        67       7979
817        66       7869
818        60       7755
819        54       7513

[819 rows x 2 columns]

In [302]:
df2.loc[ : , 'deathCnt' : 'stateDt' ]

deathCnt  decideCnt  seq   stateDt
1       24299   18174842  903  20220607
2       24279   18168670  902  20220606
3       24258   18163648  901  20220605
4       24238   18153814  900  20220604
5       24229   18141775  899  20220603
..        ...        ...  ...       ...
815        72       8086   55  20200314
816        67       7979   54  20200313
817        66       7869   53  20200312
818        60       7755   52  20200311
819        54       7513   51  20200310

[819 rows x 4 columns]

In [303]:
### df의 컬럼 이름을 변경
### ['등록일시', '총사망자', '총확진자', '게시글번호', '기준일', '기준시간', '수정일시', '누적의심자', '누적확진자']

### DataFrame은 클래스. 클래스는 변수와 함수의 모음.
적어도 다음 3개의 변수를 가지고 있다.
- columns -> ['createDT', ..... ] 배열 데이터
- index -> 0~819까지의 데이터
- values -> [
                [],
                [],
                ...
                []
                   ]

df.columns 치면
[           ] 데이터 생성

df.columns = [ ] 이렇게 치면
칼럼 바꿔주는 것
원래 컬럼 개수와 변경할 컬럼 이름 개수가 똑같아야 함.

In [304]:
df.columns

Index(['createDt', 'deathCnt', 'decideCnt', 'seq', 'stateDt', 'stateTime',
       'updateDt', 'accExamCnt', 'accDefRate'],
      dtype='object')

In [305]:
corona = df.copy()

In [306]:
corona.columns

Index(['createDt', 'deathCnt', 'decideCnt', 'seq', 'stateDt', 'stateTime',
       'updateDt', 'accExamCnt', 'accDefRate'],
      dtype='object')

In [307]:
df.columns

Index(['createDt', 'deathCnt', 'decideCnt', 'seq', 'stateDt', 'stateTime',
       'updateDt', 'accExamCnt', 'accDefRate'],
      dtype='object')

In [308]:
df2.columns

Index(['createDt', 'deathCnt', 'decideCnt', 'seq', 'stateDt', 'stateTime',
       'updateDt', 'accExamCnt', 'accDefRate'],
      dtype='object')

In [318]:
df.columns = ['등록일시', '총사망자', '총확진자', '게시글번호', '기준일', '기준시간', '수정일시', '누적의심자', '누적확진자']

In [319]:
df.head(1)

등록일시   총사망자      총확진자  게시글번호       기준일   기준시간 수정일시  \
0  2022-06-08 09:09:05.982  24305  18188200    904  20220608  00:00  NaN   

   누적의심자  누적확진자  
0    NaN    NaN

In [311]:
corona.columns

Index(['createDt', 'deathCnt', 'decideCnt', 'seq', 'stateDt', 'stateTime',
       'updateDt', 'accExamCnt', 'accDefRate'],
      dtype='object')

In [312]:
df.columns = corona.columns

In [313]:
df.head(1)


createDt  deathCnt  decideCnt  seq   stateDt stateTime  \
0  2022-06-08 09:09:05.982     24305   18188200  904  20220608     00:00   

  updateDt  accExamCnt  accDefRate  
0      NaN         NaN         NaN

In [315]:
## rename() : 컬럼의 값이나 인덱스 값을 변경할 때 사용하는 함수
## rename(columns= {dict}, index = {dict})

corona.rename(
    columns= {
        'createDt' : '등록일시',
        'deathCnt' : '총사망자',
        'decideCnt' : '총확진자',
        'seq' : '게시글번호', 
        'stateDt' : '기준일', 
        'stateTime' : '기준시간',
       'updateDt' : '수정일시', 
       'accExamCnt' : '누적의심자', 
       'accDefRate' : '누적확진자'
    }

    , inplace = True #이렇게 저장해버린다/ 왠만해선 바로 바꾸지 말 것.
)

In [320]:
df.head(1)

등록일시   총사망자      총확진자  게시글번호       기준일   기준시간 수정일시  \
0  2022-06-08 09:09:05.982  24305  18188200    904  20220608  00:00  NaN   

   누적의심자  누적확진자  
0    NaN    NaN

In [321]:
corona.head(1)

등록일시   총사망자      총확진자  게시글번호       기준일   기준시간 수정일시  \
0  2022-06-08 09:09:05.982  24305  18188200    904  20220608  00:00  NaN   

   누적의심자  누적확진자  
0    NaN    NaN

In [324]:
# df에 있는 특정 컬럼의 이름을 변경
# 누적확진자 -> 누적확진율
df.rename(
    columns = {
        '누적확진자' : '누적확진율'
    }, inplace= True #이건 리스트 한 번 확인한 다음에 추가 실행할 것
)

In [338]:
# inplace = True를 이해하기 위한 풀어쓰기 과정

class Test:
    def __init__(self, data, index = None, columns = None):
        self.data = data
        if index:
            self.index = index
        else:
            self.index = list(range(len(data)))
        if columns:
            self.columns = columns
        else:
            self.columns = list(range(len(data[0])))
    def __repr__(self):
        return str(self.data)   #repr이라는 함수를 쓰려면 문자열이어야 해서 str을 적어준 것.
    
    def add(self, value, inplace = False):
        if inplace:
            self.data = self.data + [value]
        else:
            return self.data + [value]

In [339]:
value = [
    [1,2,3],
    [7,8,9],
    [4,5,6]
]


test_class = Test(value)

In [340]:
test_class

[[1, 2, 3], [7, 8, 9], [4, 5, 6]]

In [341]:
test_class.index

[0, 1, 2]

In [342]:
test_class.columns

[0, 1, 2]

In [349]:
test_class.add(1) #단순 출력, 데이터 변하지 않음.

[[1, 2, 3], [7, 8, 9], [4, 5, 6], 1, 1, 1]

In [347]:
test_class.add(1, inplace=True) #데이터도 변함.

In [348]:
test_class

[[1, 2, 3], [7, 8, 9], [4, 5, 6], 1, 1]

In [352]:
df.head(3)

등록일시   총사망자      총확진자  게시글번호       기준일   기준시간  \
0  2022-06-08 09:09:05.982  24305  18188200    904  20220608  00:00   
1  2022-06-07 09:09:00.897  24299  18174842    903  20220607  00:00   
2  2022-06-06 09:00:06.734  24279  18168670    902  20220606  00:00   

                      수정일시  누적의심자  누적확진율  
0                      NaN    NaN    NaN  
1  2022-06-08 09:10:36.846    NaN    NaN  
2  2022-06-08 09:10:50.441    NaN    NaN

In [358]:
#등록일시를 기준으로 오름차순 정렬을 해보자.
#sort_index는 행을 기준으로 정렬한다는 의미이고
#sort_values는 값을 기준으로 정렬한다는 의미
# ctrl + space 누르면 자동완성 창이 뜸. 맥에서는 command + space.

df.sort_values('등록일시', ascending=True, inplace=True)
#ascending : 오름차순(True), 내림차순(False)을 지정. 오름차순이 기본값.

In [359]:
df.head(3)

등록일시  총사망자  총확진자  게시글번호       기준일   기준시간  \
819  2020-03-10 00:00:00.000    54  7513     51  20200310  00:00   
818  2020-03-11 00:00:00.000    60  7755     52  20200311  00:00   
817  2020-03-12 00:00:00.000    66  7869     53  20200312  00:00   

                       수정일시     누적의심자     누적확진율  
819  2021-10-07 10:30:51.51  210144.0  3.919308  
818  2021-10-07 10:30:51.51  222395.0  3.804175  
817  2021-10-07 10:30:51.51  234998.0  3.621744

In [362]:
## 데이터프레임의 인덱스를 재지정(초기화)
df.reset_index(drop = True, inplace = True) #()안을 비우면, index column이 추가되고, drop =True라고 쓰면 기존 인덱스가 초기화된다.

In [367]:
df.values

array([['2020-03-10 00:00:00.000', 54, 7513, ...,
        '2021-10-07 10:30:51.51', 210144.0, 3.919308057],
       ['2020-03-11 00:00:00.000', 60, 7755, ...,
        '2021-10-07 10:30:51.51', 222395.0, 3.804174536],
       ['2020-03-12 00:00:00.000', 66, 7869, ...,
        '2021-10-07 10:30:51.51', 234998.0, 3.621744273],
       ...,
       ['2022-06-06 09:00:06.734', 24279, 18168670, ...,
        '2022-06-08 09:10:50.441', nan, nan],
       ['2022-06-07 09:09:00.897', 24299, 18174842, ...,
        '2022-06-08 09:10:36.846', nan, nan],
       ['2022-06-08 09:09:05.982', 24305, 18188200, ..., nan, nan, nan]],
      dtype=object)

In [371]:
# 인덱스의 조건이 1이고 컬럼의 조건이 총사망자인 데이터와
# 인덱스의 조건이 0이고 컬럼의 조건이 총사망자인 데이터의 차이 -> 인덱스1의 일일 사망자
df.loc[1, '총사망자'] - df.loc[0, '총사망자']

# 위의 작업을 [인덱스 개수 - 1] 만큼 반복

# 일일사망자가 추가될 리스트를 생성
# 첫날의 데이터는 0으로 지정
_list = [0]

for i in range(len(df) -1): #함수를 모를 때 쓰는 식이 바로 for 구문..
    data = df.loc[ i+1, '총사망자' ] - df.loc[ i, '총사망자' ]
    _list.append(data)

_list

[0,
 6,
 6,
 1,
 5,
 3,
 0,
 6,
 3,
 7,
 3,
 8,
 2,
 7,
 9,
 6,
 5,
 8,
 5,
 8,
 6,
 4,
 3,
 4,
 5,
 3,
 6,
 3,
 6,
 8,
 4,
 4,
 3,
 3,
 3,
 5,
 3,
 4,
 1,
 2,
 2,
 2,
 1,
 1,
 2,
 0,
 0,
 2,
 1,
 1,
 2,
 1,
 1,
 2,
 0,
 2,
 2,
 1,
 1,
 0,
 0,
 0,
 0,
 2,
 1,
 1,
 2,
 0,
 1,
 0,
 0,
 1,
 0,
 2,
 0,
 1,
 2,
 0,
 0,
 0,
 0,
 1,
 1,
 1,
 1,
 0,
 0,
 0,
 0,
 0,
 1,
 2,
 0,
 1,
 0,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 1,
 0,
 2,
 1,
 0,
 1,
 0,
 0,
 0,
 2,
 2,
 1,
 1,
 1,
 0,
 1,
 0,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 1,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 2,
 0,
 0,
 0,
 1,
 2,
 1,
 3,
 5,
 2,
 1,
 0,
 2,
 3,
 2,
 2,
 1,
 2,
 5,
 3,
 2,
 4,
 5,
 3,
 5,
 4,
 0,
 5,
 5,
 1,
 5,
 2,
 3,
 0,
 5,
 2,
 4,
 2,
 5,
 1,
 6,
 2,
 1,
 4,
 1,
 1,
 0,
 3,
 2,
 1,
 2,
 2,
 1,
 1,
 4,
 1,
 2,
 2,
 1,
 0,
 3,
 3,
 3,
 2,
 2,
 0,
 0,
 3,
 1,
 1,
 1,
 1,
 2,
 2,
 4,
 2,
 1,
 1,
 1,
 1,
 2,
 5,
 2,
 0,
 1,
 4,
 1,


In [369]:
len(df)

820

In [372]:
len(_list)

820

In [373]:
df['일일사망자'] = _list #이퀄은 할당

In [374]:
df.head(3)

등록일시  총사망자  총확진자  게시글번호       기준일   기준시간  \
0  2020-03-10 00:00:00.000    54  7513     51  20200310  00:00   
1  2020-03-11 00:00:00.000    60  7755     52  20200311  00:00   
2  2020-03-12 00:00:00.000    66  7869     53  20200312  00:00   

                     수정일시     누적의심자     누적확진율  일일사망자  
0  2021-10-07 10:30:51.51  210144.0  3.919308      0  
1  2021-10-07 10:30:51.51  222395.0  3.804175      6  
2  2021-10-07 10:30:51.51  234998.0  3.621744      6

In [ ]:
#가장 원초적인 방법이었다..

In [375]:
df['총사망자']

0         54
1         60
2         66
3         67
4         72
       ...  
815    24238
816    24258
817    24279
818    24299
819    24305
Name: 총사망자, Length: 820, dtype: int64

In [377]:
# 스리즈 형태에 value를 인덱스 이동
# shift(n) : n의 기본값은 1
 # 2를 넣으면 두 칸씩 이동, 3을 넣으면 세 칸씩 이동, 위로 한 칸씩 올리려면 -1

df['총사망자'].shift()

0          NaN
1         54.0
2         60.0
3         66.0
4         67.0
        ...   
815    24229.0
816    24238.0
817    24258.0
818    24279.0
819    24299.0
Name: 총사망자, Length: 820, dtype: float64

In [378]:
# 결측치가 있음. 

df['총사망자'] - df['총사망자'].shift(1) #1은 기본값이므로 안 넣어도 됨.

0       NaN
1       6.0
2       6.0
3       1.0
4       5.0
       ... 
815     9.0
816    20.0
817    21.0
818    20.0
819     6.0
Name: 총사망자, Length: 820, dtype: float64

In [381]:
df['총사망자'] - df['총사망자'].shift(1).fillna(0) #결측치를 0으로 채워준다
#이때 54가 뜸.

0      54.0
1       6.0
2       6.0
3       1.0
4       5.0
       ... 
815     9.0
816    20.0
817    21.0
818    20.0
819     6.0
Name: 총사망자, Length: 820, dtype: float64

In [383]:
(df['총사망자'] - df['총사망자']).shift(1).fillna(0) #괄호 처리를 저기까지만 하는 경우.

0      0.0
1      0.0
2      0.0
3      0.0
4      0.0
      ... 
815    0.0
816    0.0
817    0.0
818    0.0
819    0.0
Name: 총사망자, Length: 820, dtype: float64

In [382]:
(df['총사망자'] - df['총사망자'].shift(1)).fillna(0) #shift까지 괄호 처리를 해줘야 해당 결측치 0 처리가 잘 됨.

0       0.0
1       6.0
2       6.0
3       1.0
4       5.0
       ... 
815     9.0
816    20.0
817    21.0
818    20.0
819     6.0
Name: 총사망자, Length: 820, dtype: float64

In [384]:
df['일일사망자2'] = (df['총사망자'] - df['총사망자'].shift(1)).fillna(0)

In [385]:
df.head(3)

등록일시  총사망자  총확진자  게시글번호       기준일   기준시간  \
0  2020-03-10 00:00:00.000    54  7513     51  20200310  00:00   
1  2020-03-11 00:00:00.000    60  7755     52  20200311  00:00   
2  2020-03-12 00:00:00.000    66  7869     53  20200312  00:00   

                     수정일시     누적의심자     누적확진율  일일사망자  일일사망자2  
0  2021-10-07 10:30:51.51  210144.0  3.919308      0     0.0  
1  2021-10-07 10:30:51.51  222395.0  3.804175      6     6.0  
2  2021-10-07 10:30:51.51  234998.0  3.621744      6     6.0

In [386]:
## 스리즈 데이터에서 데이터간의 차이를 출력하는 함수가 존재함.
## diff(n) n이 의미하는 건 shift와 마찬가지로 몇 칸 이동한 데이터와 차이가 있는가? 기본값은 당연히 1.
## diff(n) : n만큼 인덱스가 이동한 데이터와의 차이를 출력 (n = 1 기본값)
df['총사망자'].diff()

0       NaN
1       6.0
2       6.0
3       1.0
4       5.0
       ... 
815     9.0
816    20.0
817    21.0
818    20.0
819     6.0
Name: 총사망자, Length: 820, dtype: float64

In [387]:
df['일일사망자3'] = (df['총사망자'].diff().fillna(0))

In [388]:
df.head(5)

등록일시  총사망자  총확진자  게시글번호       기준일   기준시간  \
0  2020-03-10 00:00:00.000    54  7513     51  20200310  00:00   
1  2020-03-11 00:00:00.000    60  7755     52  20200311  00:00   
2  2020-03-12 00:00:00.000    66  7869     53  20200312  00:00   
3  2020-03-13 00:00:00.000    67  7979     54  20200313  00:00   
4  2020-03-14 00:00:00.000    72  8086     55  20200314  00:00   

                     수정일시     누적의심자     누적확진율  일일사망자  일일사망자2  일일사망자3  
0  2021-10-07 10:30:51.51  210144.0  3.919308      0     0.0     0.0  
1  2021-10-07 10:30:51.51  222395.0  3.804175      6     6.0     6.0  
2  2021-10-07 10:30:51.51  234998.0  3.621744      6     6.0     6.0  
3  2021-10-07 10:30:51.51  248647.0  3.458499      1     1.0     1.0  
4  2021-10-07 10:30:51.51  261335.0  3.318000      5     5.0     5.0

In [400]:
### 일일 확진자 파생변수 생성
### 일일확진자 -> shift()을 이용하여 생성
### 일일확진자2 -> diff()을 이용하여 생성

###describe() 함수를 이용하여 데이터의 이상이 있는가 확인

In [418]:
df['일일확진자'] = (df['총확진자'] - df['총확진자'].shift()).fillna(0)

In [419]:
df.head(5)

등록일시  총사망자  총확진자  게시글번호       기준일   기준시간  \
0  2020-03-10 00:00:00.000    54  7513     51  20200310  00:00   
1  2020-03-11 00:00:00.000    60  7755     52  20200311  00:00   
2  2020-03-12 00:00:00.000    66  7869     53  20200312  00:00   
3  2020-03-13 00:00:00.000    67  7979     54  20200313  00:00   
4  2020-03-14 00:00:00.000    72  8086     55  20200314  00:00   

                     수정일시     누적의심자     누적확진율  일일사망자  일일사망자2  일일사망자3  일일확진자  \
0  2021-10-07 10:30:51.51  210144.0  3.919308      0     0.0     0.0    0.0   
1  2021-10-07 10:30:51.51  222395.0  3.804175      6     6.0     6.0  242.0   
2  2021-10-07 10:30:51.51  234998.0  3.621744      6     6.0     6.0  114.0   
3  2021-10-07 10:30:51.51  248647.0  3.458499      1     1.0     1.0  110.0   
4  2021-10-07 10:30:51.51  261335.0  3.318000      5     5.0     5.0  107.0   

   일일확진자2  
0     0.0  
1   242.0  
2   114.0  
3   110.0  
4   107.0

In [420]:
df['일일확진자2'] = df['총확진자'].diff().fillna(0)

In [421]:
df.head(5)

등록일시  총사망자  총확진자  게시글번호       기준일   기준시간  \
0  2020-03-10 00:00:00.000    54  7513     51  20200310  00:00   
1  2020-03-11 00:00:00.000    60  7755     52  20200311  00:00   
2  2020-03-12 00:00:00.000    66  7869     53  20200312  00:00   
3  2020-03-13 00:00:00.000    67  7979     54  20200313  00:00   
4  2020-03-14 00:00:00.000    72  8086     55  20200314  00:00   

                     수정일시     누적의심자     누적확진율  일일사망자  일일사망자2  일일사망자3  일일확진자  \
0  2021-10-07 10:30:51.51  210144.0  3.919308      0     0.0     0.0    0.0   
1  2021-10-07 10:30:51.51  222395.0  3.804175      6     6.0     6.0  242.0   
2  2021-10-07 10:30:51.51  234998.0  3.621744      6     6.0     6.0  114.0   
3  2021-10-07 10:30:51.51  248647.0  3.458499      1     1.0     1.0  110.0   
4  2021-10-07 10:30:51.51  261335.0  3.318000      5     5.0     5.0  107.0   

   일일확진자2  
0     0.0  
1   242.0  
2   114.0  
3   110.0  
4   107.0

In [422]:
df.describe()

###describe() 함수를 이용하여 데이터의 이상이 있는가 확인

## min값이 음수가 나옴. 어떻게 찾나? 0보다 작은 데이터가 들어오면 안 되는데.

총사망자          총확진자       게시글번호           기준일         누적의심자  \
count    820.000000  8.200000e+02  820.000000  8.200000e+02  6.920000e+02   
mean    3927.832927  1.913097e+06  472.393902  2.020897e+07  7.545057e+06   
std     6164.310693  4.857030e+06  249.701095  7.118265e+03  6.172020e+06   
min       54.000000  7.513000e+03   51.000000  2.020031e+07  2.101440e+05   
25%      415.750000  2.393525e+04  256.750000  2.020100e+07  1.934309e+06   
50%     1812.500000  1.185640e+05  461.500000  2.021042e+07  6.368310e+06   
75%     3120.500000  3.979915e+05  698.250000  2.021112e+07  1.216890e+07   
max    24305.000000  1.818820e+07  904.000000  2.022061e+07  2.151807e+07   

            누적확진율       일일사망자      일일사망자2      일일사망자3          일일확진자  \
count  623.000000  820.000000  820.000000  820.000000     820.000000   
mean     1.556435   29.574390   29.574390   29.574390   22171.569512   
std      0.543014   68.530863   68.530863   68.530863   71013.854189   
min      0.902205   -2.000000   -2.000000   -2.000000    -430.000000   
25%      1.078089    2.000000    2.000000    2.000000     125.000000   
50%      1.416159    5.000000    5.000000    5.000000     642.500000   
75%      1.816009   20.000000   20.000000   20.000000    2782.750000   
max      3.919308  469.000000  469.000000  469.000000  621204.000000   

              일일확진자2  
count     820.000000  
mean    22171.569512  
std     71013.854189  
min      -430.000000  
25%       125.000000  
50%       642.500000  
75%      2782.750000  
max    621204.000000

In [424]:
# 일일사망자 데이터에 음수인 데이터만 출력
df['일일사망자'] < 0

0      False
1      False
2      False
3      False
4      False
       ...  
815    False
816    False
817    False
818    False
819    False
Name: 일일사망자, Length: 820, dtype: bool

In [432]:
flag = df['일일사망자'] < 0
df.loc[flag, ]

#인덱스 확인은 가능..

등록일시  총사망자    총확진자  게시글번호       기준일   기준시간  \
446  2021-05-30 00:00:00.000  1957  139907    497  20210530  00:00   

                       수정일시      누적의심자     누적확진율  일일사망자  일일사망자2  일일사망자3  \
446  2021-10-07 10:30:51.51  9747612.0  1.454166     -2    -2.0    -2.0   

     일일확진자  일일확진자2  
446 -430.0  -430.0

In [433]:
flag2 = df['일일확진자'] < 0
df.loc[flag2, ]

등록일시  총사망자    총확진자  게시글번호       기준일   기준시간  \
446  2021-05-30 00:00:00.000  1957  139907    497  20210530  00:00   

                       수정일시      누적의심자     누적확진율  일일사망자  일일사망자2  일일사망자3  \
446  2021-10-07 10:30:51.51  9747612.0  1.454166     -2    -2.0    -2.0   

     일일확진자  일일확진자2  
446 -430.0  -430.0

In [434]:
df.loc[444:448, ]

등록일시  총사망자    총확진자  게시글번호       기준일   기준시간  \
444  2021-05-29 00:00:00.000  1951  139427    496  20210529  00:00   
445    2021-05-30 00:00:00.0  1959  140337    498  20210531  00:00   
446  2021-05-30 00:00:00.000  1957  139907    497  20210530  00:00   
447  2021-06-01 00:00:00.000  1963  140796    499  20210601  00:00   
448  2021-06-02 00:00:00.000  1965  141473    500  20210602  00:00   

                        수정일시      누적의심자     누적확진율  일일사망자  일일사망자2  일일사망자3  \
444   2021-10-07 10:30:51.51  9733588.0  1.450884      5     5.0     5.0   
445  2021-10-14 13:48:56.821  9761156.0  1.456805      8     8.0     8.0   
446   2021-10-07 10:30:51.51  9747612.0  1.454166     -2    -2.0    -2.0   
447   2021-10-07 10:30:51.51  9798400.0  1.455890      6     6.0     6.0   
448   2021-10-07 10:30:51.51  9834348.0  1.457668      2     2.0     2.0   

     일일확진자  일일확진자2  
444  533.0   533.0  
445  910.0   910.0  
446 -430.0  -430.0  
447  889.0   889.0  
448  677.0   677.0

In [435]:
## 445번의 날짜 포맷팅이 달라서 문제구나, 싶고
## 5월은 31일까지 있음.
## 게시글번호 순서가 445, 446간 뒤바껴 있음
## 445번 데이터의 수정일시는 일주일 뒤
## 2개의 순서가 바뀌면 되겠구나!
## 그럼 순서를 어떻게 바꿀 수 있을까?
## 445번 일시를 5월 31일로 바꿔준다! - raw data 수정

df.loc[445, '등록일시']

'2021-05-30 00:00:00.0'

In [439]:
df.loc[445, '등록일시'] = '2021-05-31 00:00:00.000'

In [440]:
df.loc[444:448, ]

등록일시  총사망자    총확진자  게시글번호       기준일   기준시간  \
444  2021-05-29 00:00:00.000  1951  139427    496  20210529  00:00   
445  2021-05-31 00:00:00.000  1959  140337    498  20210531  00:00   
446  2021-05-30 00:00:00.000  1957  139907    497  20210530  00:00   
447  2021-06-01 00:00:00.000  1963  140796    499  20210601  00:00   
448  2021-06-02 00:00:00.000  1965  141473    500  20210602  00:00   

                        수정일시      누적의심자     누적확진율  일일사망자  일일사망자2  일일사망자3  \
444   2021-10-07 10:30:51.51  9733588.0  1.450884      5     5.0     5.0   
445  2021-10-14 13:48:56.821  9761156.0  1.456805      8     8.0     8.0   
446   2021-10-07 10:30:51.51  9747612.0  1.454166     -2    -2.0    -2.0   
447   2021-10-07 10:30:51.51  9798400.0  1.455890      6     6.0     6.0   
448   2021-10-07 10:30:51.51  9834348.0  1.457668      2     2.0     2.0   

     일일확진자  일일확진자2  
444  533.0   533.0  
445  910.0   910.0  
446 -430.0  -430.0  
447  889.0   889.0  
448  677.0   677.0

In [442]:
#등록일시 기준으로 오름차순 정렬
#일일확진자, 일일사망자 재계산
#여기서 멈추고 다시 2주 후에 해볼 것!
df.ascend[]

SyntaxError: invalid syntax (309446471.py, line 1)